#BE-4783 Add profile creation settings

hiilhan opened the pull request.

hiilhan committed:
add profile creation settings

hiilhan committed:
minor loop fixes

hiilhan committed:
input validation, mapping

hiilhan committed:
add testing groups option

hiilhan committed:
declare organizationId parameter

hiilhan committed:
replace org. id with name

hiilhan committed:
implement subOrganizationName

burakberkkeskin reviewed (approved):
Tested on dev environment:
- The profile created with static username password successfully. Shared with email and downloaded successfully.
- The profile created with non-auth successfully. Shared with email and downloaded successfully.

hiilhan committed:
give full name for auth types

hiilhan committed:
update undefined auth type handling

hiilhan committed:
make authType string

hiilhan committed:
add profile update tip

hiilhan committed:
fix grammar

hiilhan committed:
fix testing groups nil input, convert testing groups not found warning to error

hiilhan committed:
Merge pull request #8 from appcircleio/feature/BE-4784

#BE-4784 Add testing groups option

hiilhan committed:
Merge branch 'feature/BE-4783' into feature/BE-4921

hiilhan committed:
Merge pull request #9 from appcircleio/feature/BE-4921

#BE-4921 Add sub organization name option

hiilhan committed:
remove the multipart warning

hiilhan committed:
remove readme details and add docs reference

hiilhan committed:
distribution -> testing distribution

hiilhan committed:
fix error message

hiilhan committed:
revert raise removals

csonuryilmaz commented on README.md:
````diff
@@ -80,28 +67,43 @@ fastlane add_plugin appcircle_testing_distribution
 ```yml
   appcircle_testing_distribution(
     personalAPIToken: "$(AC_PERSONAL_API_TOKEN)",
+    subOrganizationName: "${AC_SUB_ORGANIZATION_NAME}",
````
`{` vs. `(` Which one should be?

csonuryilmaz commented on README.md:
```diff
@@ -80,28 +67,43 @@ fastlane add_plugin appcircle_testing_distribution
+      authType: "${AC_PROFILE_AUTH_TYPE}",
+      username: "${AC_PROFILE_USERNAME}",
+      password: "${AC_PROFILE_PASSWORD}",
+      testingGroupNames: "AC_PROFILE_TESTING_GROUP_NAMES"
```
Fix

csonuryilmaz commented on README.md:
```diff
@@ -80,28 +67,43 @@ fastlane add_plugin appcircle_testing_distribution
 - `message`: Your message to testers, ensuring they receive important updates and information regarding the application.
 
-### Leveraging Environment Variables
+## Docs
```
```suggestion
## Further Details
```

hiilhan committed:
log profile found

hiilhan committed:
fix auth type validation

hiilhan committed:
change env usage in the example

hiilhan committed:
update readme:
- sync with docs
- remove undefined authType warning

hiilhan committed:
update action:
- remove undefined authType option
- move app path validation to verify block
- create get_or_create_profile method
- fix error logging style
- refactor operator syntax

hiilhan committed:
improve null safety for upload service

csonuryilmaz commented on README.md:
````diff
@@ -79,29 +66,44 @@ fastlane add_plugin appcircle_testing_distribution
 
 ```yml
````
It should be `ruby` instead of yaml.

hiilhan replied:
1. Done
2. Done
3. Updated to 0.3.2-dev

csonuryilmaz commented on lib/fastlane/plugin/appcircle_testing_distribution/helper/TDUploadService.rb:
```diff
@@ -10,7 +11,7 @@ def self.upload_artifact(token:, message:, app:, dist_profile_id:)
     headers = {
       Authorization: "Bearer #{token}",
-      content_type: :multipart # multipart/form-data
+      # content_type: :multipart or multipart/form-data # Causing a type cast warning
```
❓ Are we sure about this change? Does it affect anything?

Content type might be an important argument in request headers.

hiilhan replied:
Didn't affect the tests, I will check again.

hiilhan replied:
> Auto-detects that you have a File value sends it as multipart
> (from https://github.com/rest-client/rest-client#multipart)

Current payload has a File value but added `multipart: true`, to make sure. Tested the result, request didn't change.

csonuryilmaz replied:
OK. Thanks for the research and clarification.

csonuryilmaz commented on lib/fastlane/plugin/appcircle_testing_distribution/helper/TDUploadService.rb:
```diff
@@ -2,6 +2,7 @@
 require 'uri'
 require 'json'
 require 'rest-client'
+require_relative '../helper/appcircle_testing_distribution_helper'
```
AFAIK, we do not refer to any function from this helper ruby. So, it seems to be unnecessary to "require" the file here.

csonuryilmaz commented on README.md:
```diff
@@ -79,29 +66,44 @@ fastlane add_plugin appcircle_testing_distribution
-
-Efficiently distribute test binaries or beta versions using Appcircle, featuring seamless IPA and APK distribution capabilities. Streamline your testing process with our versatile tool designed to optimize your distribution workflow. If you need support or more information, please [contact us](https://appcircle.io/contact?utm_source=fastlane&utm_medium=plugin&utm_campaign=testing_distribution)
+- [Setting Up Appcircle Testing Distribution Plugin](https://docs.appcircle.io/marketplace/fastlane/testing-distribution)
+   - [Discover Action](https://docs.appcircle.io/marketplace/fastlane/testing-distribution#discover-action)
```
Markdown indentation should be 2 spaces instead of 3 spaces for unordered list items. (markdown lint)

csonuryilmaz commented on README.md:
```diff
@@ -79,29 +66,44 @@ fastlane add_plugin appcircle_testing_distribution
-- `appPath`: Indicates the file path to the application that will be uploaded to Appcircle Testing Distribution Profile.
+- `createProfileIfNotExists` (optional): Ensures that a testing distribution profile is automatically created if it does not already exist; if the profile name already exists, the app will be uploaded to that existing profile instead.
+- `profileCreationSettings` (optional): If `createProfileIfNotExists` is `true` and a new profile being created, the profile will be configured with these settings.
+   - `authType`: Authentication type of the profile. `none`: None, `static`: Static Username and Password, `ldap`: LDAP Login, `sso`: SSO Login.
```
Markdown indentation should be 2 spaces instead of 3 spaces for unordered list items. (markdown lint)

csonuryilmaz reviewed (changes requested):
@hiilhan I have added some threads to be resolved and one question to discuss.

In addition to those, please also fix the cases below.

1️⃣ Remove obsolete images from repository.
- `images/distribution-start.png`
- `images/PAT.png`

AFAIK, we are using the CDN URL in the README. Also, they might be stale.

2️⃣ The license should be owned by Appcircle instead of an individual contributor. You can check other samples from workflow step license files.

![Screenshot 2024-12-18 at 03 25 01](https://github.com/user-attachments/assets/b948598e-39db-494d-aa94-731565daba15)

3️⃣ Update the `version.rb` file to our target version: `0.3.2`. So, our PR will be ready for release when merged.

hiilhan committed:
update redundant content

hiilhan committed:
lint readme

hiilhan committed:
change version to 0.3.2-dev

csonuryilmaz commented on fastlane/Fastfile:
```diff
@@ -4,9 +4,16 @@ Dotenv.load
 lane :test do
   appcircle_testing_distribution(
     personalAPIToken: ENV["AC_ACCESS_TOKEN"],
```
Change to `AC_PERSONAL_API_TOKEN`

csonuryilmaz commented on README.md:
```diff
@@ -65,43 +52,59 @@ To generate a Personal API Token, follow these steps:
-    createProfileIfNotExists: Boolean,
-    appPath: "$(AC_APP_PATH)",
-    message: "$(AC_MESSAGE)",
+    personalAPIToken: ENV["AC_ACCESS_TOKEN"],
```
Change to `AC_PERSONAL_API_TOKEN`

hiilhan committed:
fix naming of AC_PERSONAL_API_TOKEN

hiilhan committed:
force multipart encoding

hiilhan committed:
remove env_name support for all parameters
